### PR TITLE
containerd handle localhost and ports in registry hosts

### PR DESCRIPTION
**Pull request by chrisplo**
localhost is a valid server reference but does not have a '.'
server reference may also have a port and be a locally
resolvable hostname

### CODE REVIEW

File: pkg/image/containerd/daemon_provider.go [hunk 1]
```diff
@@ -507,10 +507,16 @@ func withMetadata(platform *platforms.Platform, ref string) (metadata []image.Ad
...
-	if len(parts) == 1 {
+	switch {
+	case len(parts) == 0:
+		return imageName
+	case len(parts) == 1:
 		return fmt.Sprintf("docker.io/library/%s", imageName)
-	} else if len(parts) > 1 && !strings.Contains(parts[0], ".") {
+	case strings.Contains(parts[0], ".") || strings.Contains(parts[0], ":"):
+		return imageName
+	case parts[0] == "localhost" || parts[0] == "127.0.0.1":
```
[the remark this answers is not in the document]
Comment (chrisplo) [reply pkg/image/containerd/daemon_provider.go:517]: i split a couple cases with same return values to keep the column width not too wide (readability?)  can adjust for style as needed

[the remark this answers is not in the document]
Comment (chrisplo) [reply pkg/image/containerd/daemon_provider.go]: as a side note, I'm using this in context of syft, and replacing master with my PR fixes the issue I saw for using localhost:<port> tagged images in containerd

could not determine source: an error occurred attempting to resolve 'localhost:32000/<my image>:0.1.0-7749': containerd: unable to parse registry reference="docker.io/localhost:32000/<my image>:0.1.0-7749": could not parse reference: docker.io/localhost:32000/<my image>:0.1.0-7749

[the remark this answers is not in the document]
Comment (kzantow) [reply pkg/image/containerd/daemon_provider.go]: But there's no case where `len(parts) == 0`. Even `strings.Split("", "/")` results in `[]string{""}`, which is len() == 1.

[the remark this answers is not in the document]
Comment (kzantow) [reply pkg/image/containerd/daemon_provider.go]: How about let's try to make this even more clear what the intent is by adding a helper function called something like `isHostname`.

Then we'd have something like:
```
switch {
case len(parts) == 1:
  return fmt.Sprintf("docker.io/library/%s", imageName)
case isHostname(parts[0]):
  return imageName
default:
  return fmt.Sprintf("docker.io/%s", imageName)
}
```

and
```
func isHostname(s string) bool {
  return s == "localhost" || strings.Contains(s, ".") || strings.Contains(s, ":")
}
```

[the remark this answers is not in the document]
Comment (chrisplo) [reply pkg/image/containerd/daemon_provider.go]: noticed we don't need to test for len > 1 as we know can't be zero and we test early for len == 1

[the remark this answers is not in the document]
Comment (chrisplo) [reply pkg/image/containerd/daemon_provider.go]: hi @kzantow, is there more I can do for this PR?

[the remark this answers is not in the document]
Comment (chrisplo) [reply pkg/image/containerd/daemon_provider.go]: hi @kzantow, it's been a month, would you mind taking another look at latest?

**Review by kzantow — APPROVED** [review]
Sorry I missed the updates here, @chrisplo! This LGTM! Thanks! 👍
